### PR TITLE
Fix chall selection

### PR DESCRIPTION
**Pull request by jdkasten**
This should enable use with the Boulder server.  I fixed a few bugs in the challenge selection and added some unittests.

### CODE REVIEW

File: letsencrypt/acme/challenges.py [hunk 1]
```diff
@@ -184,7 +184,7 @@ class Hints(jose.JSONObjectWithFields):
         """Hints for "proofOfPossession" challenge.
 
         :ivar jwk: JSON Web Key (:class:`letsencrypt.acme.jose.JWK`)
-        :ivar list certs: List of :class:`M2Crypto.X509.X509` cetificates.
+        :ivar list certs: List of :class:`M2Crypto.X509.X509` certificates.
```
Comment (kuba) [inline letsencrypt/acme/challenges.py]: they should probably be also wrapped in `ComparableX509`

File: letsencrypt/client/auth_handler.py [hunk 2]
```diff
@@ -315,24 +315,27 @@ def _challenge_factory(self, domain, path):
 def gen_challenge_path(challs, preferences, combinations):
     """Generate a plan to get authority over the identity.
 
-    .. todo:: Make sure that the challenges are feasible...
-        Example: Do you have the recovery key?
+    .. todo:: This can be possibly be rewritten to use resolved_combinations.
```
Comment (kuba) [inline letsencrypt/client/auth_handler.py]: :)

File: letsencrypt/client/auth_handler.py [hunk 3]
```diff
@@ -387,13 +395,14 @@ def _find_dumb_path(challs, preferences):
     assert len(preferences) == len(set(preferences))
 
     path = []
-    satisfied = set()
+    # This cannot be a set() because POP challenge is not currently hashable
```
Comment (kuba) [inline letsencrypt/client/auth_handler.py]: What do you mean exactly (error traceback, etc.)? I'd like to fix it. And if you don't care to wait for the fix, could you mark this line as "TODO" so it's easier to spot?

File: letsencrypt/client/tests/auth_handler_test.py [hunk 4]
```diff
@@ -513,6 +513,78 @@ def test_not_satisfied(self):
...
+        self.assertTrue(self._call(challs, prefs, None))
+
+    def test_not_supported(self):
+        challs = (acme_util.POP, acme_util.DVSNI)
+        prefs = [challenges.DVSNI]
+        combos = ((0, 1),)
+
+        self.assertRaises(errors.LetsEncryptAuthHandlerError,
+                          self._call,
+                          challs, prefs, combos)
```
Comment (kuba) [inline letsencrypt/client/tests/auth_handler_test.py]: you could probably squeeze above into two lines

**Comment by kuba** [comment]
LGTM :+1:  modulo minor comments above

File: letsencrypt/client/auth_handler.py (hunk 3 above)
Comment (kuba) [inline letsencrypt/client/auth_handler.py]: oh, unittests catch it when `set` is used, cool

Comment (kuba) [inline letsencrypt/client/auth_handler.py]: fixed in #327; could you please merge it and revert back to `set`?

Comment (jdkasten) [inline letsencrypt/client/auth_handler.py]: Given that the set is only beneficial as an optimization if the server happens to send the exact same challenge multiple times (which we are already entering an undefined area here).  I will attempt to change it back after I go over #327.

**Comment by jdkasten** [comment]
Thanks!

**Pull request merged**